### PR TITLE
Graph meter coloring & dynamic scaling!

### PR DESCRIPTION
**Pull request by eworm-de**
This work is by @Explorer09, see original pull request (https://github.com/hishamhm/htop/pull/487).
I rebased to current master to prepare this pull request.

### CODE REVIEW

**Comment by eworm-de** [comment]
This is what it can look like, really helpful if cpu usage is reflected in graph meters.

Screenshot_2020-09-16_13-25-55

File: Meter.c
Comment (BenBE) [inline Meter.c:471]: As we're not talking hundrets of stacked values, this should be negligible …

```suggestion
         if (stack1[items + 1] > 0) {
            scale = this->full / stack1[items + 1];
            for (int i = 0; i < items; i++) {
               stack1[i + 1] = stack1[i + 1] * scale;
            }
         }
```

File: Meter.c
Comment (BenBE) [inline Meter.c:453]: Some comment what those variables hold would be fine.

File: Meter.c
Comment (BenBE) [inline Meter.c:486]: Avoid this line break for better readability.

File: Meter.c
Comment (BenBE) [inline Meter.c:543]: Short, highlevel description would be fine. Also this looks awfully complicated and should probably be split into its own function.

Also some blank lines for better visual grouping might help.

File: Meter.c
Comment (BenBE) [inline Meter.c:533]: Avoid line break for better readability.

File: Meter.c
Comment (BenBE) [inline Meter.c:549]: One variable per declaration.

File: Meter.h
Comment (BenBE) [inline Meter.h:43]: No float-equal compare … ;-)

File: Meter.h
Comment (BenBE) [inline Meter.h:73]: These names are somewhat unintuitive. Care to add a comment for their usage?

File: Meter.h
Comment (BenBE) [inline Meter.h:118]: These should be checked by `configure` …

File: Meter.h
Comment (BenBE) [inline Meter.h:128]: Use `static inline` and define 3 proper versions.
Also: Does this really have to life in the header or could this be placed into `Meter.c`?

File: zfs/ZfsCompressedArcMeter.h
Comment (BenBE) [inline zfs/ZfsCompressedArcMeter.h:14]: Why is this changed to `extern`? This shouldn't be necessary for regular functions.

File: zfs/ZfsCompressedArcMeter.h
Comment (BenBE) [inline zfs/ZfsCompressedArcMeter.h:12]: Headers not sorted.

**Review by BenBE — CHANGES_REQUESTED** [review]

File: Meter.h
Comment (Explorer09) [reply Meter.h:128]: @BenBE 
If I remember correctly, htop (at least 2.x) auto-generates header files in build time, so it was not me who can decide which macros reside in the header and which only in .c file.

Also, `ilog2` is an intentionally overloaded function, and many implementations define `ilog2` as a macro. So I followed that way.

Comment (BenBE) [reply Meter.h:128]: The automatic generation of header files was removed quite a while back (with 3.0.1 (in early September). And even back when the headers were still auto-generated you had quite a bit of influence as you basically wrote the headers as specially crafted comments in the .c file.

That aside: IMHO `ilog2` should probably go either into a header for utility functions (if other modules can and should make use of it) or reside in the .c file only. As it is right now, it's only used for `Meter.c` and thus is probably suited best there implementation-wise.

On the aspect of `#define` vs `static inline` function: As this implementation has several nested trinary operators, which solely depend on the argument type, it's probably more readable if done explicitly. Also for source included from elsewhere it's nice to have some reference where it came from.

```c
static inline int ilog2u(unsigned int x) {
   return CHAR_BIT*sizeof(x)-1-__builtin_clz(x);
}
static inline int ilog2ul(unsigned long x) {
   return CHAR_BIT*sizeof(x)-1-__builtin_clzl(x);
}
static inline int ilog2ull(unsigned long long x) {
   return CHAR_BIT*sizeof(x)-1-__builtin_clzll(x);
}
```

The only implementation I could find (in a quick search) was the one from the Linux kernel in
[tools/includes/linux/log2.h line 71](https://github.com/torvalds/linux/blob/master/tools/include/linux/log2.h#L71) (which only special-cases numbers up to 4 bytes and those larger than this and apart from this handles the actual implementation as separate functions).

**Review by eworm-de — COMMENTED** [review]
Already changed in later revision...

File: zfs/ZfsCompressedArcMeter.h
Comment (eworm-de) [reply zfs/ZfsCompressedArcMeter.h]: This does not (or no longer) touch that file. Nothing to care about here, no?

**Comment by cgzones** [comment]
I'll need to take a deeper look but I noticed some graphical misalignments:
graph

File: Meter.c
Comment (BenBE) [inline Meter.c:179]: What do we gain from allocating this all in one block, instead of separating these allocations by struct field?

File: Meter.c
Comment (BenBE) [inline Meter.c:169]: Missing overflow check in calculation of allocated size. Also mixing allocations of different sizes into one makes maintaining the source unnecessarily difficult.

File: Meter.c
Comment (BenBE) [inline Meter.c:197]: Could you please elaborate this description a bit more? E.g. give a bit more context.

**Comment by eworm-de** [comment]
How to proceed from here? As this is not my work but code by @Explorer09 I would like him to take over. I hope it's a good point to (re-)start after all these rebases. :)
I invited you to my repository, so you can update the commits in this pull request's branch.

**Comment by eworm-de** [comment]
> I'll need to take a deeper look but I noticed some graphical misalignments:

For me this looks like a font issue. Actually braille characters are used to show graphical meters.
Did the graphical meters show aligned before, without this series?

File: Meter.c
Comment (BenBE) [inline Meter.c:365]: Maybe move to where the SI prefix list is?

File: Meter.c
Comment (BenBE) [inline Meter.c:370]: No need for this return

File: Meter.c
Comment (BenBE) [inline Meter.c:374]: diti.

File: Meter.c
Comment (BenBE) [inline Meter.c:380]: dito.

**Comment by eworm-de** [comment]
Just for reference... I use Terminus in my terminals. For me everything looks just fine, see screenshot from first comment.

File: Meter.c
Comment (BenBE) [inline Meter.c:367]:
```suggestion
      printw("inf");
```

**Review by BenBE — CHANGES_REQUESTED** [review]

File: Meter.c
Comment (Explorer09) [reply Meter.c:169]: @BenBE 
If the data to be allocated is from user input, I would have already considered the size overflow problem seriously. But for this case, how come would a raster  graph buffer take even a megabyte of memory space? Sorry, but I don't believe there would be a security issue here.

Second, there is no alignment problem here. C compiler handles struct alignment automatically, and misalignment can only happen if you override the alignment (via `#pragma pack` and such) or you cast a pointer in order to do an unaligned read.

Comment (BenBE) [reply Meter.c:169]: As mentioned, I'm mostly over-cautious in this regard and regard combining allocations as bad style if no clear necessity for doing so can be given.

Regarding the alignment: Just after the allocation comes code, that casts the memory buffers around. The types are fine (and alignment only gets smaller), but it's a pitfall waiting for someone to trip over.

**Comment by Explorer09** [comment]
Hi everyone reviewing this pull request,

I personally wish a redesign to be done on the graph rendering code as I think the current state of algorithm isn't ideal.
Before I have the time and motivation to write a new version of code, I wish some prerequite commits to be considered merging first.
Some commits in this PR is really about variable naming and code style, and it would be way faster to me if the following are decided.

* Whether we will use the GRAPH_NUM_RECORDS macro
* Meter.total versus Meter.full, which name is better
* The ilog2 thing - should I keep it? As you know __builtin_clz is a directive specific to GCC and Clang. But I guess there are some people who don't like compiler-specic directives anywhere.

**Comment by BenBE** [comment]
> * The ilog2 thing - should I keep it? As you know __builtin_clz is a directive specific to GCC and Clang. But I guess there are some people who don't like compiler-specic directives anywhere.

If you move that code into e.g. `Compat.[ch]` we should be fine. Only condition there is that it should yield correct results independent of the compile environment.

> * Meter.total versus Meter.full, which name is better

I find neither very intuitive …

> I personally wish a redesign to be done on the graph rendering code as I think the current state of algorithm isn't ideal.

Especially trying to understand the basic idea behind it is harder than it should be *winks at certain review comments on that subject*. What I'd wish for in the new algorithm would be some high-level explanation on the data structures (can be short and precise, no elaborate essays required) and some step-by-step walk-through as comments to follow along with the processing as it goes (partially done in recent PR revisions).

File: Meter.c
Comment (BenBE) [inline Meter.c:54]: Why copy this locally?

Comment (Explorer09) [reply Meter.c:54]: It won't matter because compiler is smart enough to optimize this one out.

**Comment by fasterit** [comment]
@eworm-de , @Explorer09: We looked at this PR during today's core team call.
There are still quality issues as mentioned during the Oct 30 discussion above.
You also mentioned that you would like to "redesign [..] the graph rendering code".

Are the any plans to fix the issue preventing a merge currently and / or is there any work being done on the redesign?

**Comment by eworm-de** [comment]
I had hoped for @Explorer09 to chime in...
I handled all the rebases and simple & obvious review comments. Not sure my understanding is deep enough for fundamental changes. Also I am unaware if the rewrite is in progress already...

**Comment by Explorer09** [comment]
@eworm-de I'm really sorry that I was too busy and involved in my paid job and didn't look at this.
I wonder, do you really love to see these changes be merged to master?

I thank you for your help anyway, but I wonder if you can help revert this d418d73258654a1f621a1f018778fac68c142071 because I think it disrupts the code in too many places.

**Comment by eworm-de** [comment]
Yes, I want to see this in master! (At the moment I handle this manually by rebasing and building a modified package, but it's an extra burden.)

Sure, will revert that commit. It caused a lot of work in the past...

**Comment by eworm-de** [comment]
Rebased and dropped the rename.

**Comment by Explorer09** [comment]
@eworm-de Excuse me, but it seems like you have made it wrong with the commit b036b9eab0374d3f18bc7a6518d09a4904756d63 ("Introduce GRAPH_NUM_RECORDS macro constant"). The meaning is different from the version I originally had, and I found it out when I rebase this branch onto the current master.
(The GRAPH_NUM_RECORDS constant is now unnecessary since 2d1042adb382aa95456beff31b40abb7196567a0 introduced METER_GRAPHDATA_SIZE which have the same meaning.)

**Comment by eworm-de** [comment]
Ups, sorry...
I will happily push your rebased code to my branch.

**Comment by Explorer09** [comment]
@eworm-de Can you please help me test my new version?
The algorithm of this is reworked almost entirely, but I cannot test it thoroughly to ensure it's bug free.
I wish this pull request be closed in favor of the new version.
https://github.com/Explorer09/htop/tree/graph-color-new-4

**Comment by BenBE** [comment]
@Explorer09 Just open a new PR for this and leave a comment regarding superseding this one.

**Comment by eworm-de** [comment]
You should have a look a #714 where progress is tracked.

**Comment by Explorer09** [comment]
@AdityaSriram09 Hello, and thank you for being interested in this "graph coloring" feature. As eworm-de mentioned, this pull request has been obsolete as I rewrote the code in #714. You can test the new one and I would like to hear your opinion. Note that the algorithm in #714 is not yet final as I plan to rewrite it again for the goal of rendering the graph as representative of the original data as possible.

**Comment by natoscott** [comment]
@Explorer09 @eworm-de would either/both of you guys be interested in mentoring a Google Summer of Code participant like @AdityaSriram09 to work with you on this effort?  If you could describe the details of those rewrite plans and your goals there, I'm sure a student would be able to assist and learn a whole lot in the process.  Please send me a note (scott . nato @ gmail . com) if interested and I'll assist with the administrative side of things.

**Comment by AdityaSriram09** [comment]
@Explorer09, thanks for the advice, I'll look into #714 and get back to you. It would be great if @Explorer09 and @eworm-de can share details of rewrite plans and goals as @natoscott suggested. Thanks in advance !

**Comment by Explorer09** [comment]
@AdityaSriram09 I'm sorry, but I planned to rewrite the algorithm myself, and I don't think you can be involved much. The best thing you can do is to study the new code of mine and see if there is anything not documented well enough (how it works, rationale or anything), and suggest improvements of that.

**Comment by natoscott** [comment]
@Explorer09 this work could benefit from others helping to move it forward with you.  Definitely a detailed write up of what needs improving in the current code, what is planned, would be extremely helpful for everyone - yourself included I think.

The original PR ([355 in hishamhm's repo](https://github.com/hishamhm/htop/pull/355) ) was opened in 2016, more than 6 years ago now.  Quite a few top-like tools have similar functionality now and it would be great to see this in htop soon.

**Comment by Explorer09** [comment]
@natoscott Would you please name examples of such functionality? I wish to compare the algorithms of how colors are determined with the one in my PR.
The reason I said I planned a rewrite is the current algorithm is not perfect. [When two or more items have the same value, the algorithm tend to give space to lower part of the graph.](https://github.com/htop-dev/htop/pull/714#issuecomment-894919999) I wish the overall position of every item be considered when allocating the colors. That's the main part I'm trying to fix.

**Comment by natoscott** [comment]
> @natoscott Would you please name examples of such functionality?

https://computingforgeeks.com/top-terminal-based-monitoring-tools-for-linux/

> I wish to compare the algorithms of how colors are determined with the one in my PR. The reason I said I planned a rewrite is the current algorithm is not perfect. [When two or more items have the same value, the algorithm tend to give space to lower part of the graph.](https://github.com/htop-dev/htop/pull/714#issuecomment-894919999) I wish the overall position of every item be considered when allocating the colors. That's the main part I'm trying to fix.

OK, thanks - are you actively working on this?  can we get a GSoC participant to assist?  they want to help and I'm sure they will learn alot from you on this project.

**Pull request closed without merging**
